File: Code/Components/DanpheEMR.ServerModel/PharmacyModels/PHRMStoreModel.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using Newtonsoft.Json;

namespace DanpheEMR.ServerModel
{
    public class PHRMStoreModel
    {
        [Key]
        public int StoreId { get; set; }
        public string Category { get; set; }
        public string SubCategory { get; set; }
        public int ParentStoreId { get; set; }
        public string Name { get; set; }
        public string StoreDescription { get; set; }
        public int PermissionId { get; set; }
        public int MaxVerificationLevel { get; set; }
        public DateTime CreatedOn { get; set; }
        public int CreatedBy { get; set; }
        public DateTime? ModifiedOn { get; set; }
        public int? ModifiedBy { get; set; }
        public Boolean IsActive { get; set; }
        public string StoreLabel { get; set; }
        public string PanNo { get; set; }
        public string Code { get; set; }
        public string Address { get; set; }
        public string ContactNo { get; set; }
        public string Email { get; set; }
        public bool UseSeparateInvoiceHeader { get; set; }
        [NotMapped]
        public List<StoreVerificationMapModel> StoreVerificationMapList { get; set; }
        [NotMapped]
        public bool IsDispensary { get => Category == "dispensary"; }

        #region Payment Modes
        public string AvailablePaymentModesJSON { get; private set; }
        public string DefaultPaymentMode { get; private set; }

        [NotMapped]
        public virtual ICollection<PaymentModesSettings> AvailablePaymentModes
        {
            get
            {
                if (string.IsNullOrWhiteSpace(AvailablePaymentModesJSON))
                {
                    if (IsDispensary)
                        return new List<PaymentModesSettings>() { new PaymentModesSettings("cash") };
                    else
                        return new List<PaymentModesSettings>();
                }
                return JsonConvert.DeserializeObject<List<PaymentModesSettings>>(AvailablePaymentModesJSON);
            }
        }

        public void AddPaymentMode(string PaymentModeName, bool IsRemarksMandatory = false)
        {
            if (!IsDispensary) throw new InvalidOperationException("Payment Mode feature is only available for Disensary.");
            if (string.IsNullOrWhiteSpace(PaymentModeName)) throw new ArgumentException($"'{nameof(PaymentModeName)}' cannot be null or whitespace.", nameof(PaymentModeName));

            AvailablePaymentModes.Add(new PaymentModesSettings(PaymentModeName, IsRemarksMandatory));
            AvailablePaymentModesJSON = JsonConvert.SerializeObject(AvailablePaymentModes);
        }
        public void RemovePaymentMode(string PaymentModeName)
        {
            if (!IsDispensary) throw new InvalidOperationException("Payment Mode feature is only available for Disensary.");
            if (string.IsNullOrWhiteSpace(PaymentModeName)) throw new ArgumentException($"'{nameof(PaymentModeName)}' cannot be null or whitespace.", nameof(PaymentModeName));

            var paymentModeToRemove = AvailablePaymentModes.FirstOrDefault(a => a.PaymentModeName == PaymentModeName);
            if (paymentModeToRemove == null) throw new ArgumentNullException($"'{nameof(PaymentModeName)}' was not found in Available Payment Modes.", nameof(PaymentModeName));
            AvailablePaymentModes.Remove(paymentModeToRemove);
            AvailablePaymentModesJSON = JsonConvert.SerializeObject(AvailablePaymentModes);
        }
        public void SetDefaultPaymentMode(string PaymentMode)
        {
            if (!IsDispensary) throw new InvalidOperationException("Payment Mode feature is only available for Disensary.");
            if (string.IsNullOrWhiteSpace(PaymentMode)) throw new ArgumentException($"'{nameof(PaymentMode)}' cannot be null or whitespace.", nameof(PaymentMode));
            if (!AvailablePaymentModes.Any(a => a.PaymentModeName == PaymentMode)) throw new ArgumentException($"'{nameof(PaymentMode)}' cannot be default if it is not in Available Payment Modes.", nameof(PaymentMode));

            DefaultPaymentMode = PaymentMode;
        }
        #endregion Payment Modes

        #region Inventory Receipt Numbers Settings
        public int? INV_GRGroupId { get; set; }
        public int? INV_POGroupId { get; set; }
        public int? INV_PRGroupId { get; set; }
        public int? INV_ReqDisGroupId { get; set; }
        public int? INV_RFQGroupId { get; set; }
        public string INV_ReceiptDisplayName { get; set; }
        public string INV_ReceiptNoCode { get; set; }
        #endregion
    }

    public class PaymentModesSettings
    {
        public string PaymentModeName { get; set; }
        public bool IsRemarksMandatory { get; set; }

        public PaymentModesSettings(string paymentModeName, bool isRemarksMandatory = false)
        {
            PaymentModeName = paymentModeName ?? throw new ArgumentNullException(nameof(paymentModeName));
            IsRemarksMandatory = isRemarksMandatory;
        }
    }
}

File: Code/Websites/DanpheEMR/wwwroot/DanpheApp/src/app/pharmacy/shared/phrm-store.model.ts
import { FormGroup, Validators, FormBuilder } from '@angular/forms'
import { StoreVerificationMapModel } from '../../settings-new/shared/store-role-map.model';


export class StoreBillHeader {
  PanNo: string = '';
  StoreLabel: string = '';
  Code: string = '';
  Address: string = '';
  ContactNo: string = '';
  Email: string = '';
}

export class PHRMStoreModel extends StoreBillHeader {
  public StoreId: number = 0;
  public ParentStoreId: number = 0;
  public Name: string = "";
  public ParentName: string = "";
  public StoreDescription: string = "";
  public MaxVerificationLevel: number = 0;
  public PermissionId: number;
  public CreatedBy: number = 0;
  public CreatedOn: string = "";
  public ModifiedBy: number = 0;
  public ModifiedOn: string = "";
  public IsActive: boolean = true;
  public Category: string;
  public SubCategory: string;
  public StoreVerificationMapList: Array<StoreVerificationMapModel> = new Array<StoreVerificationMapModel>();
  public StoreValidator: FormGroup = null;
  UseSeparateInvoiceHeader: boolean = false;
  AvailablePaymentModes: PaymentModeSettings[] = [];
  DefaultPaymentMode: string;

  // Receipt Variables
  INV_GRGroupId : number;
  INV_POGroupId : number;
  INV_PRGroupId : number;
  INV_ReqDisGroupId : number;
  INV_RFQGroupId : number;
  INV_ReceiptDisplayName : string;
  INV_ReceiptNoCode : string;
  constructor() {
    super();
    var _formBuilder = new FormBuilder();
    this.StoreValidator = _formBuilder.group({
      'Name': ['', Validators.compose([Validators.required, Validators.maxLength(50)])],
      'Email': ['', Validators.pattern('^[A-Z0-9a-z._%+-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,64}$')],
      'ContactNo': ['', [Validators.pattern('^[a-zA-Z0-9_@./#)(&+-]+$'), Validators.maxLength(20)]],
      'StoreLabel': ['']
    });
  }

  public IsDirty(fieldName): boolean {
    if (fieldName == undefined)
      return this.StoreValidator.dirty;
    else
      return this.StoreValidator.controls[fieldName].dirty;
  }

  public IsValid(): boolean { if (this.StoreValidator.valid) { return true; } else { return false; } } public IsValidCheck(fieldName, validator): boolean {
    if (fieldName == undefined) {
      return this.StoreValidator.valid;
    }
    else
      return !(this.StoreValidator.hasError(validator, fieldName));
  }
}

class PaymentModeSettings {
  PaymentModeName: string;
  IsRemarksMandatory: boolean = false;
}

